gateway-js/src/outOfBandReporter.ts

```diff
@@ -0,0 +1,108 @@
+import { fetch, Response, Request } from 'apollo-server-env';
+import { ApiMonitoringReport, ErrorCode } from './__generated__/graphqlTypes';
+
+
+// Magic /* GraphQL */ comment below is for codegen, do not remove
+export const OUT_OF_BAND_REPORTER_QUERY = /* GraphQL */`#graphql
+  mutation OOBReport($input: APIMonitoringReport) {
+    reportError(report: $input)
+  }
+`;
+
+const { name, version } = require('../package.json');
+
+export class OutOfBandReporter {
+  endpoint: string | null;
+
+  constructor() {
+    this.endpoint = process.env.APOLLO_OUT_OF_BAND_REPORTER_ENDPOINT || null;
+  }
+
+  async submitOutOfBandReport({
+    error,
+    request,
+    response,
+    startedAt,
+    endedAt,
+    tags,
+    fetcher,
+  }: {
+    error: Error;
+    request: Request;
+    response?: Response;
+    startedAt: Date;
+    endedAt: Date;
+    tags?: string[];
+    fetcher: typeof fetch;
+  }) {
+
+    // don't send report if the endpoint url is not configured
+    if (!this.endpoint) {
+      return;
+    }
+
+    let errorCode = ErrorCode.Other;
+
+    // some possible error situations to check against
+    if (response?.status && [400, 413, 422].includes(response?.status)) {
+      errorCode = ErrorCode.InvalidBody;
+    }
+    else if (response?.status && [408, 504].includes(response?.status)) {
+      errorCode = ErrorCode.Timeout;
+    }
+    else if (!response || [502, 503].includes(response?.status)) {
+      errorCode = ErrorCode.ConnectionFailed;
+    }
+
+    let oobResponse: Response;
+    let responseBody: string;
+    try{
+      responseBody = await response?.json();
```

Comment:
This is a great call out, since the api is expecting the response body as string, it actually makes more sense just to use body.text() instead of decoding the json!

Reply:
I like the outcome here!